### DISCUSSION TITLE
[Controller] Symfony 6, container and autowiring.

### DISCUSSION THREAD

**Discussion by lugosium** [symfony/symfony#44628] [Q&A] [answered]
Before Symfony 6, it was possible to not have autoconfigure and autowire for our controllers with:

```
App\Controller\DefaultController:
        arguments:
            - '@some_bundle_service'
        calls:
            - [setContainer, ['@service_container']]
        tags:
            - 'controller.service_arguments'
```

With Symfony 6 it's not working anymore, i guess because of the service locator. Even if i add the tag `container.service_subscriber`, i need to enable autowire.

Is there a way or do we need autowire for controller now?

EDIT TLTR: 

explanation: https://github.com/symfony/symfony/discussions/44628#discussioncomment-1807168
working example: https://github.com/symfony/symfony/discussions/44628#discussioncomment-1963592

**Comment by stof** [comment, marked as answer]
If you want to use the `container.service_subscriber` tag and have the corresponding managed service locator injected, you should set a call to `setContainer` without passing the `service_container` service as argument (which injects the full container) but using the `Psr\Container\ContainerInterface` id instead (for which the RegisterServiceSubscribersPass adds a binding).
And I'm quite sure that this was also the case on Symfony 5 (the difference is that some services were public in SF 5 and became private in SF 6, making them unavailable when injecting `service_container`)

**Reply by lugosium** [reply]
Thanks for your quick answer.

I do not want to use the `container.service_subscriber` tag specially, but trying to upgrade to version 6 from 5.4 and i got issues with the container because of the private services like you said.

I don't understand how to retrieve the corresponding managed service locator and inject it.
Is it possible in YAML? It's some logic with a service id suffixed by some class? 
I don't get it, and debug:container don't helped me.

With the `Psr\Container\ContainerInterface` container injected (which id is `parameter_bag` in my install) i also got the same issues with private services. But maybe i didn't understand well.

Sorry, i 🏊‍♂️

**Reply by lugosium** [reply]
I looked at the code of `RegisterServiceSubscribersPass`, my controller is working if i:

`- [setContainer, ['@.service_locator.jIxfAsi.App\Controller\DefaultController']]`

But i retrieved the id with a dump inside `RegisterServiceSubscribersPass`...

**Reply by lugosium** [reply]
Anyway, I don't think it's possible as those service locator are internal, they are not meant to be used so i will enable autowire and autoconfigure until i find another solution.

**Reply by lugosium** [reply]
Thank you @stof for your help, i will keep an eye on discussions/issues.

**Reply by trsteel88** [reply]
@lugosium, did you find a solution for this? I'm interested in doing the same for a legacy application and I need access to private services.

**Reply by lugosium** [reply]
Nop, i enabled autowire.

**Reply by stof** [reply]
What is necessary for wiring AbstractController child classes is to wire them as service subscribers with the `container.service_subscriber`, because AbstractController is a service subscriber.
This can be done in 2 ways:
- adding the `container.service_subscriber` tag yourselves on the service definition
- turning on autoconfiguration (which will add that tag automatically)

**Reply by lugosium** [reply]
So, after all your messages today, i have tested again and confirm that it's working without autowiring and autoconfigure. 

Thank you @stof, @ahundiak and @trsteel88.

Discussion updated with links.

**Comment by ahundiak** [comment]
As a rule, "does not work" is not a very useful description.  There is no reason why your original service definition should not work in Symfony 6.  Here is a working example of the same thing: https://github.com/cerad/s6bundle/blob/master/src-my-bundle/Resources/config/services.yaml

If I had to guess I would suspect that routing is the real issue.  Maybe check debug:router.  Maybe post the relevant portion of your error message.

**Reply by trsteel88** [reply]
This doesn't work for private services though. I forked your repository and made a single change: https://github.com/trsteel88/s6bundle/commit/eae7286eb95d504280790c4627b4835182330971

You will now see 'You cannot use the "renderView" method if the Twig Bundle is not available. Try running "composer require symfony/twig-bundle".'

Even though 'twig' is part of the Subscribed Services.

**Reply by stof** [reply]
please read my answer above. Your service definition should not inject `@service_container`, as that injects the main container, which does not allow getting private services). That's not how the AbstractController is meant to be wired.

**Reply by ahundiak** [reply]
@trsteel88 The original question was about configuration.  For whatever reason the OP was reluctant to share any error messages.  As you discovered the twig service was indeed changed from public to private between 5.4 and 6.0.  Which is really kind of special when you think about it.

I personally stopped using the AbstractController some years ago.  Too much magical configuration for me.

**Reply by ahundiak** [reply]
@stof Even if you add the tag (manually or with autoconfgure) you still need autowire enabled as well.

**Reply by stof** [reply]
no you don't. But if you don't enable autowiring, you need to configure the call `[setContainer, ['@Psr\Container\ContainerInterface']]` (as autowiring won't add it for you), as explained in my answer.

**Reply by ahundiak** [reply]
Okay.  So this works at least for accessing the private twig service:
```
# services.yaml
    my_bundle.controller.my_controller_2:
        class: MyBundle\Controller\MyController2
        calls:
            - [setContainer, ['@Psr\Container\ContainerInterface']]
        tags:
            - 'controller.service_arguments'
            - 'container.service_subscriber'
```
Still seems a bit mysterious to me.  And it appears that the original poster tried this and still had issues.  Thanks.
